src/job.rs

```diff
@@ -253,285 +252,32 @@ impl Job {
 // sent across threads.
 unsafe impl Send for Job {}
 
-#[derive(Debug)]
-pub struct JobQueue {
-    sentinel: NonNull<Job>,
-    tail: NonNull<Job>,
-    len: u32,
-}
-
-impl Default for JobQueue {
-    fn default() -> Self {
-        let root = Box::leak(Box::new(Job {
-            stack: NonNull::dangling(),
-            harness: |_, _, _| (),
-            prev: Cell::new(None),
-            fut_or_next: Cell::new(None),
-        }))
-        .into();
-
-        Self {
-            sentinel: root,
-            tail: root,
-            len: 0,
-        }
-    }
-}
-
-impl Drop for JobQueue {
-    fn drop(&mut self) {
-        // SAFETY:
-        // `self.sentinel` never gets written over, so it contains the original
-        // `leak`ed `Box` that gets allocated in `JobQueue::default`.
-        unsafe {
-            drop(Box::from_raw(self.sentinel.as_ptr()));
-        }
-    }
-}
+#[derive(Debug, Default)]
+pub struct JobQueue(VecDeque<NonNull<Job>>);
 
 impl JobQueue {
-    pub fn len(&self) -> u32 {
-        self.len
+    pub fn len(&self) -> usize {
+        self.0.len()
     }
 
     /// SAFETY:
     /// Any `Job` pushed onto the queue should alive at least until it gets
     /// popped.
     pub unsafe fn push_back<T>(&mut self, job: &Job<T>) {
-        // SAFETY:
-        // The tail can either be the root `Box::leak`ed in the default
-        // constructor or a `Job` that has been pushed previously and which is
-        // still alive.
-        let current_tail = unsafe { self.tail.as_ref() };
-        // SAFETY:
-        // This effectively casts the `Job`'s `fut_or_next` from `Future<T>` to
-        // `Future<()>` which casts the `Future`'s `Box<T>` to a `Box<()>`.
-        //
-        // This box will not be access until the pointer gets passed in the
-        // `harness` where it gets cast back to `T`.
         let next_tail = unsafe { &*(job as *const Job<T> as *const Job) };
-
-        current_tail
-            .fut_or_next
-            .set(Some(NonNull::from(next_tail).cast()));
-        next_tail.prev.set(Some(current_tail.into()));
-
-        self.len += 1;
-
-        self.tail = next_tail.into();
+        self.0.push_back(NonNull::from(next_tail).cast());
```

Comment:
This can actually be cast directly.

```suggestion
        self.0.push_back(NonNull::from(job).cast());
```